src/_guides/language/effective-dart/design.md

```diff
@@ -1005,90 +1089,254 @@ var names = people.map((Person person) {
 });
 {% endprettify %}
 
+In rare cases, the surrounding context is not precise enough to provide a type
+for one or more of the function's parameters. In those cases, you may need to
+annotate.
+
 
-### AVOID annotating with `dynamic` when not required.
+### PREFER annotating with `dynamic` even when not required.
 
-{% comment %}
-update-for-dart-2.0
-https://github.com/dart-lang/site-www/pull/480#discussion_r157355593
-{% endcomment %}
+When you omit a type annotation, *most* of the time, Dart infers a correct
+static type for you. That word "most" is the alarming part. Sometimes, inference
+fails and silently uses `dynamic` instead. If you didn't realize that inference
+failed, you now have code that *looks* statically safe but actually has no
+type checking whatsoever.
```

Comment:
This first paragraph seems to say the opposite of what the title says. It seems to be saying "annotate, or else you'll get dynamic and you'll be sorry". It becomes clearer later what you mean, but it's still a bit of a shock.

I like the words _That word “most” is the alarming part. Sometimes, inference fails and silently uses dynamic instead. If you didn’t realize that inference failed, you now have code that looks statically safe but actually has no type checking whatsoever._ but they don't seem to belong here.

I also worry about "even when not required" (in the heading) being ambiguous. It could be interpreted as "prefer annotating with dynamic over annotating with more specific types". (Also, isn't `dynamic` worse, performance-wise? It seems like we might want to mention that somewhere, or link to somewhere that discusses it.)

How about something like this instead:

### PREFER annotating with `dynamic` when you want no type checking.

When you omit a type annotation, Dart usually infers a correct static type for you. But a casual reader of your code who sees an annotation is missing has no way of knowing if you intended it to be dynamic, expected inference to fill in some other type, or simply forgot to write the annotation.

When `dynamic` is the type you want, using it explicitly makes your intent clear.

Reply:
Took a different stab at improving this.